File: Lab_7/julia_codes/julia.cpp
#import <iostream>
#include <fstream>
#include <cmath>
#include <complex>
#include <cstring>


double calcz(std::complex<double> z, std::complex<double> c)
{
  long nitmax = 1000;
  long nit = 1;
  while( abs(z) < 10.0 && nit < nitmax)
    z = z * z + c;
  double ratio = 255.0  * nit / nitmax;
  return ratio;
}

void julia_loop(double ***julia, long im_width, long im_height, double xwidth, double yheight, double xmin, double ymin, long nitmax)
{
  int nit = 0;
  double zabsmax = 10.0;
  //complex<double> z;
  std::complex<double> c(-0.1,0.65);
  std::vector<double> julia(im_width * im_height, 0.0);
  for (int ix=0; ix < im_width; ++ix)
    {
      for (int iy=0; iy < im_height; ++iy)
	{
	  nit = 0;
	  // Map pixel position to a point in the complex plane
	  std::complex<double> z((1.0 * ix) / im_width * xwidth + xmin, (1.0 * iy) / im_height * yheight + ymin);
	  // Do the iterations
	  julia[ix*im_width+iy] = calcz(z,c,zabsmax);
	}
    }
  return toPythonList(julia);
}

int main(int arc, char **argv)
{
  // define the variables then call the function over all pixels
  julia_loop(&julia, im_midth, im_height, width, yheight, xmin, ymin, nitmax);
  // write out julia to a file that then is read by juliaplot.py (see zip package)
}

File: Lab_7/julia_codes/julia.py
#/usr/bin/env python
#
# shows the Julia set (see wikipedia)
import sys
import math
import cmath
import random
import numpy as np

def calcz(z,c,zabsmax):
    '''
    calculates the recursive function z=z^2 + c; and counts how many iterations are needed to
    reach the maximum nitmax or reach the maximum for z (zabsmax). 
    The ratio (iterations/maxiterations) [range 0..1] times 255 is returned; this 
    allows to use these return values as a color scale between 0...255.
    '''
    nit = 0
    nitmax = 1000
    while abs(z) < zabsmax and nit < nitmax:
        z = z**2 + c
        nit += 1
        ratio = (float(nit) / nitmax) #* 255.0
    return ratio
        

def julia_loop(im_width, im_height, xwidth, yheight, xmin, ymin, nitmax):
    '''
    main loop that calculates recursive function for every pixel
    '''
    print("Calculate the 2D plane...")
    zabsmax = 10.0
    c = complex(-0.1,0.65)
    julia = np.zeros((im_width, im_height))
    for ix in range(im_width):
        for iy in range(im_height):
            nit = 0
            # Map pixel position to a point in the complex plane
            z = complex(float(ix) / im_width * xwidth + xmin,
                        float(iy) / im_height * yheight + ymin)
            # Do the iterations
            julia[ix][iy] = calcz(z,c,zabsmax)
    return julia

    
if __name__ == '__main__':
    if len(sys.argv)>1:
        file = sys.argv[1]
    else:
        file = 'juliadata.txt'

    print("Julia set fractal generator\n")
    im_width = 1000
    im_height = 1000
    xmin,xmax = -0.5, 0.5
    xwidth = xmax-xmin
    ymin, ymax = -0.5, 0.5
    yheight = ymax - ymin
    nitmax = 1000
    zabsmax = 10.0
    title="Julia set fractal generator"
    julia = julia_loop(im_width, im_height, xwidth, yheight, xmin, ymin, nitmax)
    with open(file,'w') as f:
        f.write(str(im_width)+'\n')
        f.write(str(im_height)+'\n')
        f.write(str(xmin)+'\n')
        f.write(str(xmax)+'\n')
        f.write(str(xwidth)+'\n')
        f.write(str(ymin)+'\n')
        f.write(str(ymax)+'\n')
        f.write(str(yheight)+'\n')
        #f.write(str(nitmax)+'\n')
        #f.write(str(zabsmax)+'\n')
        for i in julia:
            for j in i:
                f.write(str(j)+'\t')
        f.write('\n')
        

    
